File: 2020/php/src/Aoc/Days/Day10.php
<?php

namespace Ppx17\Aoc2020\Aoc\Days;

use Illuminate\Support\Collection;

class Day10 extends AbstractDay
{
    private Collection $adapters;
    private Collection $adaptersByKey;

    public function dayNumber(): int
    {
        return 10;
    }

    public function setUp(): void
    {
        $this->adapters = collect($this->getInputLines())->map(fn($x) => (int)$x)->sort()->values();
        $this->adaptersByKey = $this->adapters->flip();
    }

    public function part1(): string
    {
        $differences = [];
        $last = 0;
        $this->adapters->each(function ($adapter) use (&$last, &$differences) {
            $diff = $adapter - $last;
            $last = $adapter;
            $differences[$diff] ??= 0;
            $differences[$diff]++;
        });
        $differences[3]++; // account for the device adapter

        return $differences[1] * $differences[3];
    }

    public function part2(): string
    {
        $routes = [$this->adapters->count() - 1 => 1]; // last adapter can only reach device
        for ($index = $this->adapters->count() - 2; $index >= 0; $index--) {
            $routes[$index] = Collection::times(3)
                ->map(fn($x) => $this->adapters[$index] + $x)
                ->map(fn($nextAdapter) => $routes[$this->adaptersByKey->get($nextAdapter)] ?? 0)
                ->sum();
        }

        return Collection::times(3)
            ->filter(fn($x) => isset($this->adaptersByKey[$x]))
            ->map(fn($x) => $routes[$x - 1])
            ->sum();
    }
}

File: 2020/julia/src/Day10.jl
module Day10
include("Aoc.jl")
using .Aoc

adapters = sort(parse.(Int64, Aoc.input_lines(10)))

function part1()
    diffs = map(t -> t[1] - t[2], zip(adapters, vcat([0], adapters[1:end-1])))
    count(x -> x == 1, diffs) * (count(x -> x == 3, diffs) + 1)
end

function part2()
    routes = Dict{Int64,Int64}(length(adapters) => 1)
    a_map = Dict{Int64,Int64}()
    [a_map[v] = i for (i, v) in enumerate(adapters)]
    for i in Iterators.reverse(1:length(adapters)-1)
        routes[i] = sum(x -> get(routes, get(a_map, get(adapters, i, nothing) + x, 0), 0), 1:3)
    end
    sum(x -> get(routes, x, 0),1:3)
end
end